### PR TITLE
Feat/4-pipeline-J

### PR DESCRIPTION
**Pull request by GangBean**
## Overview
- implements BaseTrainer class and test code
- implements metric methods: precision, recall, map, ndcg

## Change Log
-

## To Reviewer
-

## Issue Tags
- See also: #4

### CODE REVIEW

File: metric.py
Comment (twndus) [inline metric.py:1]: It would be better to remove annotated lines.

File: metric.py [hunk 1]
```diff
@@ -1,11 +1,146 @@
...
+
+    return ndcg_sum / num_users
+
+def _dcg_at_k(user_actual: np.ndarray, user_predicted: np.ndarray, k: int) -> float:
+    dcg_sum: float = sum([(1.0 / log2(i+1)) for i in range(1, k+1) if user_actual[i-1] == user_predicted[i-1]])
+
+    return dcg_sum
+
+def _idcg_at_k(k:int) -> float:
+    idcg_sum: float = sum([(1 / log2(i+1)) for i in range(1, k+1)])
```
Comment (twndus) [inline metric.py]: In my opinion, IDCG has to depend on the size of user_actual as well.

File: trainers/base_trainer.py [hunk 2]
```diff
@@ -1,29 +1,155 @@
...
+        self.args: DictConfig = args
+        self.device: torch.device = self._device(self.args.device)
+        self.model: Module = self._model(self.args.model).to(self.device)
+        self.optimizer: Optimizer = self._optimizer(self.args.optimizer, self.model, self.args.lr)
+        self.loss: BCELoss = self._loss(self.args.loss)
 
-    @abstractmethod
-    def run(self):
-        pass
+    def _device(self, device_name: str) -> torch.device:
```
Comment (twndus) [inline trainers/base_trainer.py:22]: GOOD

File: trainers/base_trainer.py [hunk 3]
```diff
@@ -1,29 +1,155 @@
...
+            return torch.device(device_name.lower())
+        else:
+            logger.error(f"Not supported device: {device_name}")
+            return torch.device('cpu')
 
-    @abstractmethod
-    def train(self):
-        pass
+    def _model(self, model_name: str) -> Module:
+        if model_name.lower() in ('test',):
```
Comment (twndus) [inline trainers/base_trainer.py:30]: After First implementation, we need to change this part to an implemented model.

File: trainers/base_trainer.py [hunk 4]
```diff
@@ -1,29 +1,155 @@
...
-    @abstractmethod
     def inference(self):
         pass
 
-    @abstractmethod
     def load_best_model(self):
-        pass
\ No newline at end of file
+        logger.info(f"[Trainer] Load best model...")
+        self.model.load_state_dict(torch.load(f'{self.best_model_dir}/best_model.pt'))
```
Comment (twndus) [inline trainers/base_trainer.py:112]: The loaded model should be stored in a variable.

File: trainers/base_trainer.py [hunk 5]
```diff
@@ -1,29 +1,155 @@
...
+
+        #     valid_loss += batch_loss.item()
+
+        # valid_loss /= len(valid_dataloader)
+
+        # valid_precision_at_k, valid_recall_at_k, valid_map_at_k, valid_ndcg_at_k = self.evaluate()
+
+        # return valid_loss, valid_precision_at_k, valid_recall_at_k, valid_map_at_k, valid_ndcg_at_k
+    
+    def evaluate(self, k: int=20) -> tuple[float]:
```
Comment (twndus) [inline trainers/base_trainer.py:107]: How about adding an explicit decorator for the abstract method?

**Review by twndus — COMMENTED** [review]
In my view, the IDCG function appears to need adjustment.

**Review by twndus — COMMENTED** [review]
LGTM!
